<a href="https://colab.research.google.com/github/programeralebrije/MachineLearningNYtaxi/blob/main/new_york_taxi_fare_predictions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [31]:
# Jovian Commit Essentials
# Please retain and execute this cell without modifying the contents for `jovian.commit` to work
!pip install jovian --upgrade -q
import jovian
jovian.set_project('new-york-taxi-fare-predictions')
jovian.set_colab_id('1YQCyIaycmlv9Op-piXXnhOgUDYZAiKXU')

# New York Taxi Fare Predictions 

In [ ]:
!pip install jovian --upgrade --quiet

In [ ]:
import jovian

In [ ]:
# Execute this to save new versions of the notebook
jovian.commit(project="new-york-taxi-fare-predictions")

[jovian] Detected Colab notebook...
[jovian] Please enter your API key ( from https://jovian.ai/ ):
API KEY: [jovian] Uploading colab notebook to Jovian...
Committed successfully! https://jovian.ai/programeralebrije/new-york-taxi-fare-predictions


'https://jovian.ai/programeralebrije/new-york-taxi-fare-predictions'

### Outline of the project
1. Downlaod the dataset
2. Explore and analyze the dataset
3. Prepare the dataset for ML training 
4. Train Hardcoded and baseline models 
5. Make predictions and submit to Kaggle
6. Perform feature engeneering 
7. Train and evaluate different models 
8. Tune Hyperparameters for the best models 
9. Train on a GPU with the entire dataset 
10. Document and publish the project online 

# 1. Downlad the dataset 

Steps:
- Install required libraries
- Download data from Kaggle
- View dataset files 
- Load trianing set with Pandas
- Load test set with Pandas

# Install required libraries

In [ ]:
!pip install jovian opendatasets pandas numpy scikit-learn xgboost --quiet

In [ ]:
import jovian 

In [ ]:
jovian.commit()

[jovian] Detected Colab notebook...
[jovian] Uploading colab notebook to Jovian...
Committed successfully! https://jovian.ai/programeralebrije/new-york-taxi-fare-predictions


'https://jovian.ai/programeralebrije/new-york-taxi-fare-predictions'

# Download data from Kaggle 

In [ ]:
import opendatasets as od

In [ ]:
dataset_url = 'https://www.kaggle.com/c/new-york-city-taxi-fare-prediction'

In [27]:
od.download(dataset_url)

Please provide your Kaggle credentials to download this dataset. Learn more: http://bit.ly/kaggle-creds
Your Kaggle username: Your Kaggle Key: ··········


100%|██████████| 1.56G/1.56G [00:11<00:00, 142MB/s]



Extracting archive ./new-york-city-taxi-fare-prediction/new-york-city-taxi-fare-prediction.zip to ./new-york-city-taxi-fare-prediction


In [28]:
data_dir = './new-york-city-taxi-fare-prediction'

##View Dataset Files

In [29]:
# List of files with size
!ls -lh {data_dir}

total 5.4G
-rw-r--r-- 1 root root  486 Jan 24 21:08 GCP-Coupons-Instructions.rtf
-rw-r--r-- 1 root root 336K Jan 24 21:08 sample_submission.csv
-rw-r--r-- 1 root root 960K Jan 24 21:08 test.csv
-rw-r--r-- 1 root root 5.4G Jan 24 21:11 train.csv


In [30]:
!wc -1 {data_dir}/train.csv

wc: invalid option -- '1'
Try 'wc --help' for more information.


In [32]:
# Training set
!head {data_dir}/train.csv

key,fare_amount,pickup_datetime,pickup_longitude,pickup_latitude,dropoff_longitude,dropoff_latitude,passenger_count
2009-06-15 17:26:21.0000001,4.5,2009-06-15 17:26:21 UTC,-73.844311,40.721319,-73.84161,40.712278,1
2010-01-05 16:52:16.0000002,16.9,2010-01-05 16:52:16 UTC,-74.016048,40.711303,-73.979268,40.782004,1
2011-08-18 00:35:00.00000049,5.7,2011-08-18 00:35:00 UTC,-73.982738,40.76127,-73.991242,40.750562,2
2012-04-21 04:30:42.0000001,7.7,2012-04-21 04:30:42 UTC,-73.98713,40.733143,-73.991567,40.758092,1
2010-03-09 07:51:00.000000135,5.3,2010-03-09 07:51:00 UTC,-73.968095,40.768008,-73.956655,40.783762,1
2011-01-06 09:50:45.0000002,12.1,2011-01-06 09:50:45 UTC,-74.000964,40.73163,-73.972892,40.758233,1
2012-11-20 20:35:00.0000001,7.5,2012-11-20 20:35:00 UTC,-73.980002,40.751662,-73.973802,40.764842,1
2012-01-04 17:22:00.00000081,16.5,2012-01-04 17:22:00 UTC,-73.9513,40.774138,-73.990095,40.751048,1
2012-12-03 13:10:00.000000125,9,2012-12-03 13:10:00 UTC,-74.006462,40.726713,-73.99

In [33]:
# Test set
!head {data_dir}/test.csv

key,pickup_datetime,pickup_longitude,pickup_latitude,dropoff_longitude,dropoff_latitude,passenger_count
2015-01-27 13:08:24.0000002,2015-01-27 13:08:24 UTC,-73.973320007324219,40.7638053894043,-73.981430053710938,40.74383544921875,1
2015-01-27 13:08:24.0000003,2015-01-27 13:08:24 UTC,-73.986862182617188,40.719383239746094,-73.998886108398438,40.739200592041016,1
2011-10-08 11:53:44.0000002,2011-10-08 11:53:44 UTC,-73.982524,40.75126,-73.979654,40.746139,1
2012-12-01 21:12:12.0000002,2012-12-01 21:12:12 UTC,-73.98116,40.767807,-73.990448,40.751635,1
2012-12-01 21:12:12.0000003,2012-12-01 21:12:12 UTC,-73.966046,40.789775,-73.988565,40.744427,1
2012-12-01 21:12:12.0000005,2012-12-01 21:12:12 UTC,-73.960983,40.765547,-73.979177,40.740053,1
2011-10-06 12:10:20.0000001,2011-10-06 12:10:20 UTC,-73.949013,40.773204,-73.959622,40.770893,1
2011-10-06 12:10:20.0000003,2011-10-06 12:10:20 UTC,-73.777282,40.646636,-73.985083,40.759368,1
2011-10-06 12:10:20.0000002,2011-10-06 12:10:20 UTC,-74.01409

Test Data Set is missing Fare amout, this set is not be use for predictions .

In [34]:
# Sample submission file
!head {data_dir}/sample_submission.csv

key,fare_amount
2015-01-27 13:08:24.0000002,11.35
2015-01-27 13:08:24.0000003,11.35
2011-10-08 11:53:44.0000002,11.35
2012-12-01 21:12:12.0000002,11.35
2012-12-01 21:12:12.0000003,11.35
2012-12-01 21:12:12.0000005,11.35
2011-10-06 12:10:20.0000001,11.35
2011-10-06 12:10:20.0000003,11.35
2011-10-06 12:10:20.0000002,11.35


In [35]:
# No. of lines in training set
!wc -l {data_dir}/train.csv

55423856 ./new-york-city-taxi-fare-prediction/train.csv


In [36]:
# No. of lines in test set
!wc -l {data_dir}/test.csv

9914 ./new-york-city-taxi-fare-prediction/test.csv


In [37]:
# No. of lines in submission file
!wc -l {data_dir}/sample_submission.csv

9915 ./new-york-city-taxi-fare-prediction/sample_submission.csv


**# No. of lines in training set**

In [38]:
!wc -l {data_dir}/train.csv

55423856 ./new-york-city-taxi-fare-prediction/train.csv


**# No. of lines in submission file**

In [39]:
!wc -l {data_dir}/test.csv

9914 ./new-york-city-taxi-fare-prediction/test.csv


**# No. of lines in test set**

In [40]:
!wc -l {data_dir}/test.csv

9914 ./new-york-city-taxi-fare-prediction/test.csv


Observations:

* This is a supervised learning 
regression problem
* Training data is 5.5 GB in size
* Training data has 5.5 million rows
* Test set is much smaller (< 10,000 rows)
* The training set has 8 columns:
   * key (a unique identifier)
   * fare_amount (target column)
   * pickup_datetime
   * pickup_longitude
   * pickup_latitude
   * dropoff_longitude
   * dropoff_latitude
   * passenger_count
* The test set has all columns except the target column fare_amount.
* The submission file should contain the key and fare_amount for each test sample.

In [41]:
jovian.commit()

[jovian] Detected Colab notebook...
[jovian] Uploading colab notebook to Jovian...
Committed successfully! https://jovian.ai/programeralebrije/new-york-taxi-fare-predictions


'https://jovian.ai/programeralebrije/new-york-taxi-fare-predictions'

##  1.- Loading Training Set

Loading the entire dataset into Pandas is going to be slow, so we can use the following optimizations:

 * Ignore the key column
 * Parse pickup datetime while loading data
 * Specify data types for other columns
   * float32 for geo coordinates
   * float32 for fare amount
   * uint8 for passenger count
* Work with a 1% sample of the data (~500k rows)
We can apply these optimizations while using pd.read_csv

In [42]:
import pandas as pd
import random

In [49]:
sample_frac = 0.03

In [44]:
selected_cols = 'fare_amount,pickup_datetime,pickup_longitude,pickup_latitude,dropoff_longitude,dropoff_latitude,passenger_count'.split(',')
dtypes = {
    'fare_amount': 'float32',
    'pickup_longitude': 'float32',
    'pickup_latitude': 'float32',
    'dropoff_longitude': 'float32',
    'passenger_count': 'float32'
}

In [45]:
selected_cols

['fare_amount',
 'pickup_datetime',
 'pickup_longitude',
 'pickup_latitude',
 'dropoff_longitude',
 'dropoff_latitude',
 'passenger_count']

In [46]:
import random

In [47]:
random.random(
    
)

0.025010755222666936

In [50]:

dtypes = {
    'fare_amount': 'float32',
    'pickup_longitude': 'float32',
    'pickup_latitude': 'float32',
    'dropoff_longitude': 'float32',
    'passenger_count': 'float32'
}

def skip_row(row_idx):
    if row_idx == 0:
        return False
    return random.random() > sample_frac

random.seed(42)
df = pd.read_csv(data_dir+"/train.csv", 
                 usecols=selected_cols, 
                 dtype=dtypes, 
                 parse_dates=['pickup_datetime'], 
                 skiprows=skip_row)

CPU times: user 5min 19s, sys: 1.04 s, total: 5min 19s
Wall time: 5min 19s

In [51]:
df

fare_amount  ... passenger_count
0               16.9  ...             1.0
1                8.9  ...             2.0
2                4.1  ...             2.0
3                4.0  ...             1.0
4               10.0  ...             1.0
...              ...  ...             ...
1661726         14.5  ...             1.0
1661727          7.5  ...             1.0
1661728          6.0  ...             1.0
1661729          8.0  ...             6.0
1661730         10.5  ...             1.0

[1661731 rows x 7 columns]

In [52]:
test_df = pd.read_csv(data_dir+'/test.csv', dtype=dtypes, parse_dates=['pickup_datetime'])

In [53]:
test_df

key  ... passenger_count
0     2015-01-27 13:08:24.0000002  ...             1.0
1     2015-01-27 13:08:24.0000003  ...             1.0
2     2011-10-08 11:53:44.0000002  ...             1.0
3     2012-12-01 21:12:12.0000002  ...             1.0
4     2012-12-01 21:12:12.0000003  ...             1.0
...                           ...  ...             ...
9909  2015-05-10 12:37:51.0000002  ...             6.0
9910  2015-01-12 17:05:51.0000001  ...             6.0
9911  2015-04-19 20:44:15.0000001  ...             6.0
9912  2015-01-31 01:05:19.0000005  ...             6.0
9913  2015-01-18 14:06:23.0000006  ...             6.0

[9914 rows x 7 columns]

In [54]:
jovian.commit()

[jovian] Detected Colab notebook...
[jovian] Uploading colab notebook to Jovian...
Committed successfully! https://jovian.ai/programeralebrije/new-york-taxi-fare-predictions


'https://jovian.ai/programeralebrije/new-york-taxi-fare-predictions'

## 2.- Explore the Dataset

* Basic info about training set
* Basic info about test set
* Exploratory data analysis & visualization
* Ask & answer questions

## Training Set


In [55]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1661731 entries, 0 to 1661730
Data columns (total 7 columns):
 #   Column             Non-Null Count    Dtype              
---  ------             --------------    -----              
 0   fare_amount        1661731 non-null  float32            
 1   pickup_datetime    1661731 non-null  datetime64[ns, UTC]
 2   pickup_longitude   1661731 non-null  float32            
 3   pickup_latitude    1661731 non-null  float32            
 4   dropoff_longitude  1661717 non-null  float32            
 5   dropoff_latitude   1661717 non-null  float64            
 6   passenger_count    1661731 non-null  float32            
dtypes: datetime64[ns, UTC](1), float32(5), float64(1)
memory usage: 57.1 MB


In [56]:
df.describe()

fare_amount  pickup_longitude  ...  dropoff_latitude  passenger_count
count  1.661731e+06      1.661731e+06  ...      1.661717e+06     1.661731e+06
mean   1.134016e+01     -7.126932e+01  ...      3.991955e+01     1.685489e+00
std    9.830639e+00      1.274979e+01  ...      1.029647e+01     1.334168e+00
min   -3.000000e+02     -3.439245e+03  ...     -3.474112e+03     0.000000e+00
25%    6.000000e+00     -7.399203e+01  ...      4.073399e+01     1.000000e+00
50%    8.500000e+00     -7.398180e+01  ...      4.075313e+01     1.000000e+00
75%    1.250000e+01     -7.396710e+01  ...      4.076806e+01     2.000000e+00
max    1.097480e+03      3.442185e+03  ...      3.351403e+03     2.080000e+02

[8 rows x 6 columns]

In [57]:
df.pickup_datetime.min(), df.pickup_datetime.max()

(Timestamp('2009-01-01 00:11:46+0000', tz='UTC'),
 Timestamp('2015-06-30 23:59:54+0000', tz='UTC'))

Observations about training data:

- 1,661,731 rows, as expected
- No missing data (in the sample)
- `fare_amount` ranges from \$-300.0 to \$1097.0 
- `passenger_count` ranges from 0 to 208 
- Dates range from 1st Jan 2009 to 30th June 2015
- The dataset takes up ~57 MB of space in the RAM

We may need to deal with outliers and data entry errors before we train our model.

# Test Set

In [58]:
test_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9914 entries, 0 to 9913
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype              
---  ------             --------------  -----              
 0   key                9914 non-null   object             
 1   pickup_datetime    9914 non-null   datetime64[ns, UTC]
 2   pickup_longitude   9914 non-null   float32            
 3   pickup_latitude    9914 non-null   float32            
 4   dropoff_longitude  9914 non-null   float32            
 5   dropoff_latitude   9914 non-null   float64            
 6   passenger_count    9914 non-null   float32            
dtypes: datetime64[ns, UTC](1), float32(4), float64(1), object(1)
memory usage: 387.4+ KB


In [59]:
test_df.describe()

pickup_longitude  pickup_latitude  ...  dropoff_latitude  passenger_count
count       9914.000000      9914.000000  ...       9914.000000      9914.000000
mean         -73.976181        40.750954  ...         40.751743         1.671273
std            0.042799         0.033542  ...          0.035435         1.278756
min          -74.252190        40.573143  ...         40.568973         1.000000
25%          -73.992500        40.736125  ...         40.735254         1.000000
50%          -73.982327        40.753052  ...         40.754065         1.000000
75%          -73.968012        40.767113  ...         40.768757         2.000000
max          -72.986534        41.709557  ...         41.696683         6.000000

[8 rows x 5 columns]

In [60]:
test_df.pickup_datetime.min(), test_df.pickup_datetime.max()

(Timestamp('2009-01-01 11:04:24+0000', tz='UTC'),
 Timestamp('2015-06-30 20:03:50+0000', tz='UTC'))

Some observations about the test set:

- 9914 rows of data
- No missing values
- No obvious data entry errors
- 1 to 6 passengers (we can limit training data to this range)
- Latitudes lie between 40 and 42
- Longitudes lie between -75 and -72
- Pickup dates range from Jan 1st 2009 to Jun  30th 2015 (same as training set)

We can use the ranges of the test set to drop outliers/invalid data from the training set.

In [61]:
jovian.commit()

[jovian] Detected Colab notebook...
[jovian] Uploading colab notebook to Jovian...
Committed successfully! https://jovian.ai/programeralebrije/new-york-taxi-fare-predictions


'https://jovian.ai/programeralebrije/new-york-taxi-fare-predictions'